### PR TITLE
Dhis2 sso

### PR DESCRIPTION
**Pull request by madoleary**
## Self proof reading checklist

- [ ] Did I run Pronto and fixed all warnings (pronto run -c origin/dev)
- [x] Make sure all naming and code will remain understandable in 6 month by someone new to the project or by you.
- [ x] Did I test the right thing?
- [ x] Did I test corner cases, non happy path (defensive testing)?
- [ ] Check that I used the ad-hoc patterns and created files accordingly (Presenters, Services, Search object, Form object,...)
- [ ] Check code efficiency with record intensive project
- [ ] Update documentation,readme accordingly

Thanks!

### CODE REVIEW

File: app/controllers/oauth/oauth_controller.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+require 'net/http'
+require 'uri'
+
+class Oauth::OauthController < Devise::OmniauthCallbacksController
+  def callback
+    # https://sandbox.bluesquare.org/uaa/oauth/authorize?client_id=[client_id]&response_type=code
+
+    program = Program.find(params["program_id"])
```
Comment (mestachs) [inline app/controllers/oauth/oauth_controller.rb:8]: we should probably test
* accessing another program 
* a non existing program

File: spec/controllers/oauth/oauth_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      it "should redirect to home" do
+        stub_request(:post, program.project_anchor.project.dhis2_url + "/uaa/oauth/token").
+          with(body: {"code"=>code, "grant_type"=>"authorization_code"}).
+          to_return(status: 200, body: "{\"access_token\":\"#{access_token}\",\"token_type\":\"bearer\",\"expires_in\":24526,\"scope\":\"ALL\"}")
+
+        stub_request(:get, program.project_anchor.project.dhis2_url + "/api/me").
+          with(headers: {"Authorization"=>"Bearer #{access_token}"}).
+          to_return(status: 200, body: "{\"id\":\"#{user.dhis2_user_ref}\"}")
+
+        @request.env["devise.mapping"] = Devise.mappings[:user]
```
Comment (mestachs) [inline spec/controllers/oauth/oauth_controller_spec.rb:31]: would extract as "def setup_devise_scope" and perhaps call in a before block

File: app/models/user.rb [hunk 3]
```diff
@@ -39,6 +41,8 @@ class User < ApplicationRecord
                       message: "Sorry, restricted signup" },
             if:     :env_dev?
 
+  validates :dhis2_user_ref, uniqueness: true, allow_blank: true
```
Comment (mestachs) [inline app/models/user.rb:44]: uniqueness is on dhis2_user_ref and program_id (2 user can have accidentaly the same id in 2 dhis2)
